### DISCUSSION TITLE
How to impement plastic strain-softening with DP/MC model + quick question on dilatant angle

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#17613] [Q&A Modules: Solid mechanics] [answered]
Hi MOOSE user,

Q1
I've just started to implement DP/MC (nonlinear) to my mine simulation.
I'm currently trying to work on strain-softening, in theory, coh and/or friction angle should be a function of the accumulated plastic strain, is there a way to implement this with moose.

so far I got DP/CM to work, but yet cannot implement softening. is there any suggestion?

Note that, we do not want to use cosserat/capped model at this stage.
-----------------------------------------------------------------------------------------------
Q2
Also another quick question regarding dilatant angle: when I implement initial_stress into the model, it gives out an error during the excavation step when I set the dilatant angle to be relatively low, for example, less than 10 degree, but the simulation works fine with it is around 20 degree. is this because the geometry is too soft and cannot handle the initial stress?

```
  [./ini_xx]
    type = ParsedFunction
    value = '-2*9.81*2700*(1000-z)'
  [../]
  [./ini_yy]
    type = ParsedFunction
    value = '-2*9.81*2700*(1000-z)'
  [../]
  [./ini_zz]
    type = ParsedFunction
    value = '-9.81*2700*(1000-z)'
  [../]
```

```
++++++++++++++++++++++++
Checking the derivatives
++++++++++++++++++++++++
Debug Parameters are as follows
stress = 


*** ERROR ***
The debug parameters have the wrong size

```

Thank you very much.
Kind regards,
Traiwit

@WilkAndy

**Comment by WilkAndy** [comment, marked as answer]
Well done for getting plasticity to work with your model.  Honestly, with caving as well it is a very difficult numerical problem.  To implement strain softening, use another `TensorMechanicsHardening` object, such as [TensorMechanicsHardeningCubic](https://mooseframework.inl.gov/source/userobjects/TensorMechanicsHardeningCubic.html) or `Exponential` or `CutExponential`.

You are almost definitely correct about the small angle not being able to support the initial stress.

**Reply by WilkAndy** [reply]
By the way, although these hardening/softening functions haven't been documented in the "new" system (actually years and years old), there is documentation at, for instance

test/tests/mohr_coulomb/hard_cubic.xlsx
test/tests/mohr_coulomb/hard1_2.xlsx

and other ".xlsx" files in the tensor_mechanics directory

**Reply by Traiwit** [reply]
Hi @WilkAndy,

Thank you! this defs helps a lot! I really wish MOOSE documentation is friendlier for new users.

just a quick question, what would be the value of 'internal parameter' in this case? is it plastic stain, or do I have to manually define it? I see the example defines it as shown below, not sure if it related.

```
  [./mc_int_auxk]
    type = MaterialStdVectorAux
    index = 0
    property = plastic_internal_parameter
    variable = mc_int
  [../]
```
Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
It's kind of my fault that the doco is not great, sorry.  When i initially developed all these plasticity models, MOOSE was using a different doco system, and although the doco is quite good, i haven't had the time to translate it across.

You don't need to define the internal param - your plasticity model (MC/DP) will do that for you.   Look at, eg
```
./test/tests/capped_mohr_coulomb/capped_mohr_coulomb.pdf
```

There is stuff on the "old wiki" but i can't get to that now.

**Reply by Traiwit** [reply]
Thank you @WilkAndy! I kinda get the idea now, I'll let you know if it works on my mine simulation  

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
Hi @WilkAndy, softening can be added to my simulation with no problem.

I have a quick question here: what is 'plastic_internal_parameter', how does moose calculate this parameter, I guess it's associated with plastic_strain, but couldn't find its definition anywhere.

not sure if it's the 'increment in total strain' from the docco from oldwiki (https://ec2-18-237-134-137.us-west-2.compute.amazonaws.com/wiki/PhysicsModules/TensorMechanics/ReturnMap/)

Kind regards,
Traiwit

**Reply by WilkAndy** [reply]
Generally, the "plastic_internal_parameter" is called q in the documentation.  For instance, about halfway down the MohrCoulomb doco you can see

```
The hardening law is defined through
h = −1 . so that q = qold + γ, so that q is the accumulated sum of γ over all "time steps".
```

The general case is written about 1/3rd of the way down the 2nd page of "Plasticity and the Return-Mapping Algorithm":
```
the increment in plastic strain is defined to be
pNαα ε ̇ = ∑ γ r . ij ij
α=1 This is the flow rule for plastic strain.  Similarly, the increment in the internal variables is
Nαα q ̇ = − ∑ γ h .
aa α=1
These are the flow rules for the internal parameters.
```
(the fonts are screwed up here - you can see the original on the webpage).

So, to figure out how an "internal variable" increases at a timestep, you use "h", and that relates the increase to "gamma".  In the MC case, \dot{q} = gamma.  Then, to figure out "gamma", you use the flow rule.  In the MC case this is epsilon_plastic = gamma * df/dsigma = \dot{q} * df/dsigma.   This gives physical meaning to the term "plastic strain".  In the MC case, df/dsigma is a horrible function that you can work out using the "Identities" webpage, but the "internal variable" is just telling you the magnitude of this plastic strain.

Roughly speaking, for MC, the size of the plastic strain and hence the "internal variable" is just what you'd intuitively think of as the plastic shear strain.

**Reply by Traiwit** [reply]
perfect! That makes sense!

Thanks again @WilkAndy 

my next step would be either adding discontinuity (fault) into the model **OR** moving to explicit method to improve the computation time (we try to replicate Abaqus explicit solver for mining simulation)
